### PR TITLE
ENH: Use container-based infrastructure

### PR DESCRIPTION
**Pull request by Eric89GXL**
Let's see if we can speed up the build system.

### CODE REVIEW

**Comment by bollu** [comment]
@Eric89GXL what kind of containers are you talking about? do you want to use pre-baked Docker containers so we only need to pull the latest Vispy?

**Comment by Eric89GXL** [comment]
We could perhaps do that eventually, but for now the idea is just to have the builds start and run faster:

http://docs.travis-ci.com/user/migrating-from-legacy/

**Comment by QuLogic** [comment]
So many fixes, you ran out of x's...

FYI, you can also place `addons` in the `env` matrix so that you don't need to install the full deps on the minimal test.

File: .travis.yml [hunk 1]
```diff
@@ -76,14 +96,9 @@ before_install:
...
     # Install numpy, flake
-    - if [ "${PYTHON}" != "2.7" ]; then
-        conda create -n testenv --yes --quiet pip python=$PYTHON > ${REDIRECT_TO};
-        source activate testenv > ${REDIRECT_TO};
-        conda install --yes --quiet numpy nose pytest > ${REDIRECT_TO};
-      else
-        travis_retry sudo apt-get -qq -y install python-numpy python-setuptools > ${REDIRECT_TO};
-        pip install -q pytest > ${REDIRECT_TO};
-      fi;
+    - conda create -n testenv --yes --quiet pip python=$PYTHON > ${REDIRECT_TO};
```
Comment (QuLogic) [inline .travis.yml]: If it's `--quiet`, what's the use of redirecting?

**Comment by Eric89GXL** [comment]
Yeah I am still tweaking :)

I didn't know that about the `env` matrix, that would be useful indeed

File: .travis.yml (hunk 1 above)
Comment (Eric89GXL) [inline .travis.yml]: it still has output

File: .travis.yml [hunk 2]
```diff
@@ -19,36 +52,23 @@ env:
...
-    - PYTHON=2.7 DEPS=full TEST=examples  # test examples
-    - PYTHON=3.4 DEPS=full TEST=examples
-
-    - PYTHON=3.4 DEPS=minimal TEST=extra  # test file sizes, style, line endings
+    - PYTHON=3.4 DEPS=full TEST=examples  # test examples
+    - PYTHON=3.4 DEPS=minimal TEST=standard  # also tests file sizes, style, line endings
 
 
 before_install:
     - REDIRECT_TO=/dev/stdout  # change to /dev/null to silence Travis
```
Comment (QuLogic) [inline .travis.yml]: With Travis' folding, I never really saw the point in this.

File: .travis.yml [hunk 3]
```diff
@@ -150,38 +136,32 @@ install:
...
         if [ "${PYTHON}" != "2.6" ]; then
-          cd ~;
-          travis_retry sudo apt-get -qq -y install mercurial libdbus-1-dev libgl1-mesa-dev libglu1-mesa-dev libpulse-dev libx11-dev libxcursor-dev libxext-dev libxi-dev libxinerama-dev libxrandr-dev libxss-dev libxt-dev libxv-dev libxxf86vm-dev libasound2-dev libts-dev libudev-dev > ${REDIRECT_TO};
           wget -q http://www.libsdl.org/release/SDL2-2.0.3.tar.gz;
           tar xzf SDL2-2.0.3.tar.gz &> ${REDIRECT_TO};
           cd SDL2-2.0.3;
           mkdir build;
           cd build;
-          ../configure --disable-audio --disable-render --disable-joystick --disable-haptic --disable-power --disable-cpuinfo &> ${REDIRECT_TO};
+          ../configure --prefix=$HOME --disable-audio --disable-render --disable-joystick --disable-haptic --disable-power --disable-cpuinfo &> ${REDIRECT_TO};
```
Comment (QuLogic) [inline .travis.yml]: Does `$HOME/.local` work? (with possibly pre-existing search paths?)

File: .travis.yml (hunk 2 above)
Comment (Eric89GXL) [inline .travis.yml]: There was a span of time (a week or a month?) where the folding didn't work properly, so it was necessary back then. Now it's probably not necessary, but also isn't doing too much harm, so we'll probably remove it eventually.

File: .travis.yml (hunk 3 above)
Comment (Eric89GXL) [inline .travis.yml]: I've tried setting LD_LIBRARY_PATH, that doesn't work. We used to use standard install directories and use `sudo ldconfig`, but we can't do that anymore.

**Comment by Eric89GXL** [comment]
@QuLogic the matrix does seem to speed up a few of the builds quite a bit, thanks for that tip. Feel free to have another look and see if it seems optimized to you.

**Comment by Eric89GXL** [comment]
@campagnola @almarklein this one is ready for merge, this speeds up our builds a bit

**Comment by almarklein** [comment]
Out of interest, what is the speed gain? Anyway, the `.travis` seems a bit cleaner. LGTM

**Comment by Eric89GXL** [comment]
Examples take about 7 min instead of about 14, primary builds about 9 instead of about 16

**Comment by Eric89GXL** [comment]
feel free to merge if you're happy, one set of eyes is probably enough on this one

**Comment by almarklein** [comment]
You asked @QuLogic to have another look above. Is he happy? :)

**Comment by almarklein** [comment]
(I find `.travis` files not the easiest to read)

**Comment by Eric89GXL** [comment]
We can always do another revision if there are additional ideas :)

File: .travis.yml [hunk 4]
```diff
@@ -1,54 +1,78 @@
...
-    - PYTHON=3.4 DEPS=full TEST=examples
-
-    - PYTHON=3.4 DEPS=minimal TEST=extra  # test file sizes, style, line endings
+    - env: PYTHON=2.6 DEPS=full TEST=standard   # functionality
+      addons:
+        apt:
+          sources:
+          packages:
+            - *mesa_apt
+    - env: PYTHON=2.7 DEPS=full TEST=standard IPYTHON="=2" # NUMPY="=1.7"
```
Comment (QuLogic) [inline .travis.yml]: Any reason to keep the `NUMPY` bit here?

Comment (Eric89GXL) [inline .travis.yml]: yeah, I want to put it back, but doing so breaks a file that I don't want to change until #928 is merged :)

**Comment by QuLogic** [comment]
Yea, seems fine to me.

**Comment by campagnola** [comment]
Looks good to me. Any idea what all the `_bootstrap.py:2150: ImportWarning: sys.meta_path is empty` is about?

**Comment by Eric89GXL** [comment]
I've seen that warning before, I think it's some weird Py3k thing from
another package.

**Comment by QuLogic** [comment]
#555 ?

**Comment by Eric89GXL** [comment]
Yeah, that one had it. That issue suggests it's from IPython. I think I've seen it before in other contexts, too.

**Comment by bollu** [comment]
Just for the record, [from the Travis blog](http://blog.travis-ci.com/2014-12-17-faster-builds-with-container-based-infrastructure/):

>Can I provide my own Docker containers?

>The build environments we're currently using on our Docker-based setup provide the same services, programming languages and tools offered by our legacy stack. ...

>We're not there yet in allowing you to bring your own, but it's something we have in mind for the future.

I think once it is allowed, we could roll our own `Docker` images?

**Comment by Eric89GXL** [comment]
Yeah probably, but I wouldn't hold my breath. They actually currently allow caching of directories, too, but only for paid users. The custom images will probably also go to paid users first, too, if I had to guess.

**Pull request merged**